### ISSUE TITLE
Make React a peer dependency

### ISSUE THREAD

**Issue by schickling** [smooch/smooch-js#245]
As we're using React 15 this plugin breaks our build because it's requiring React 14.

According to [this answer](http://stackoverflow.com/a/30454133/2418739) you should make React a peer dependency. Could you please release a new version with this in mind?

**Comment by lemieux** [comment]
Good point, I'll have to test if everything is alright (like does it work on React 15). I will most likely have time to  tackle that next week.

**Comment by schickling** [comment]
That would be brilliant. We cannot use this plugin because of this unfortunately.

**Comment by lemieux** [comment]
Are you simply installing it and doing `require('smooch')` ? If so it will need a bit of thinking around this since we publish the same bundle we publish on the CDN which contains React in it directly. 

In the meantime, feel free to fork this repo and tweak the dependencies to your needs.

**Comment by joshwcomeau** [comment]
I looked into this a bit today, because I'm also in the position where I'd rather not include React 0.14 when my app already runs React 15.

Typically, NPM modules aren't bundled; they're transpiled to ES5 and then served as a collection of individual files. As far as I know, this is the only way to specify a peer dependency without bundling it, or relying on it being set as a global variable.

I've dealt with this in the past by simply using the Babel CLI to transpile and copy the files to a `lib` folder, in addition to the UMD build in `dist`. My package.json has a `build:umd` task and a `build:lib` task, and `build` executes both of them.

Smooch-js is a bit trickier because there are other assets involved; an MP3 and a couple images. These are depending on Webpack to be bundled. 

You'd need to:

* Move React and ReactDOM from dependencies to both peerDependencies, as well as devDependencies (peer is needed to throw warnings when included without React, dev is needed for building locally).
* Add a grunt task to do the "module" build, using the Babel CLI for transpilation.
* Add a new webpack config that bundles _just_ the modules that include `require` statements for images/sounds. I'm actually not sure about this step; there may be a more elegant way around this

It's quite a bit of trouble, but I think there are some really good reasons to do it. In addition to cutting the filesize down from ~450kb to ~320kb, it makes it much easier for developers to experiment with the codebase.

I'm happy to take a shot at it if it's deemed to be worthwhile :)

**Comment by lemieux** [comment]
Hey @joshwcomeau, thanks for this, that represent pretty much what I had in mind for this. I'm sorry I didn't start working on this sooner as I said, but I had to give a hand to release a new feature in the backend and this was pushed to the next sprint. It's actually the very next thing I'll do after releasing the new feature this week.

We already deal with that kind of stuff with smooch-core (https://github.com/smooch/smooch-core-js), but smooch-js is a bit more complex as you found out yourself. Also, smooch-core came later and was built with this idea in mind from the start. 

Main thing to not forget with smooch-js, we at least need to support : 
- npm users
- bower users
- browserify users
- webpack users

Also, our webpack config injects the version as a global variable (nothing impossible to fix though).

I 100% agree the current way of doing it isn't ideal or optimal, but this was mainly made this way to be usable for most of our users. Trust me, I pushed often to have time to do it in a smarter way (and you I have it :) )

So here's my general plan to tackle this : 
- continue to generate a bundled version of the widget for the CDN and for bower users
- do the package.json changes as you stated
- create a new release script for npm that will change the version in a dedicated file and transpile source code to ES5 (would most likely end up in `lib/` in the npm distribution).
- change all `import` statement to use relative paths to allow browserify users to use transpiled files

For the media files, I was thinking of making them point to the CDN by default, but make it configurable when you call `Smooch.init` if you want to use a local version instead. That should prevent problems with non-webpack users.

Also, keep in my that the upgrade to React 15 will be out of scope for these changes and the code will still ask for React 0.14.7 in devDependencies and peerDependencies. Yay if it still works without any changes :), otherwise I'll make sure we have it on our roadmap.

In short, stay tuned, I should make some progress on this over the next two weeks :)

**Comment by joshwcomeau** [comment]
Thanks for the quick reply!

Not a problem at all :) I can imagine it's tricky supporting all these different environments. I also realize that most users probably just drop a `<script>` tag in, so I totally see why it wouldn't be top of the priorities list.

Your plan sounds great! I agree media files hosted by CDN would be easiest for this use-case; good idea to have an override in the `.init` method.

I've upgraded a couple projects from 0.14 to 15 (even an SVG-heavy project) and I've had no problems; I think there's an excellent chance that no changes will be required.

Thanks again! This project is awesome.

**Comment by lemieux** [comment]
Great, I'm still in the 2 weeks timeframe I gave :) I was stuck (again!) on another feature for the past two weeks, but I'm starting to work on this today! I didn't forget you :)

**Comment by lemieux** [comment]
Hey guys, version 2.7.0 is out and should now have React and others as peer dependencies and shouldn't bundle everything anymore!

**Comment by joshwcomeau** [comment]
Sweet! Thanks =D
